### DISCUSSION TITLE
if is there any way to store data in a global variable to share between workflow?

### DISCUSSION THREAD

**Discussion by mircotamburini** [elsa-workflows/elsa-core#2866] [Q&A] [answered]
for example any way to access asp.net distributedcache via javascript? other method?

**Comment by sfmskywalker** [comment]
One way is through a singleton service that holds a dictionary for example.
Or indeed, through a distributed cache,

You can easily extend the JavaScript function library from your own application as explained [here](https://elsa-workflows.github.io/elsa-core/docs/next/extensibility/extensibility-javascript).

For example, you could add the following functions to read & write from & to a distributed cache (untested code, I did not look at the actual API of the distributed cache service):

```csharp
public class AddCacheMethodsToJavaScript : INotificationHandler<EvaluatingJavaScriptExpression>
{
    private readonly IDistributedCache _cache;
    
    public AddCacheMethodsToJavaScript(IDistributedCache cache)
    {
       _cache = cache;
    }

    public Task Handle(EvaluatingJavaScriptExpression notification, CancellationToken cancellationToken)
    {
        var engine = notification.Engine;
        engine.SetValue("setCacheValue", (Action<string, object>) ((key, value) => _cache.Add(key, value));
        engine.SetValue("getCacheValue", (Func<string, object>) (key => _cache.Get(key));
        return Task.CompletedTask;
    }
}
```

You could even add an object to JavaScript that contains methods to read and write from and to the cache if you want to keep your JS vocabulary tidy:

```csharp
public class MyCacheService
{
    private readonly IDistributedCache _cache;

    public MyCacheService(IDistributedCache cache)
    {
        _cache = cache;
    }

    public void Set(string key, object value) => _cache.Add(key, value);
    public object Get(string key) => _cache.Get(key, value);
}

public class AddCacheMethodsToJavaScript : INotificationHandler<EvaluatingJavaScriptExpression>
{
    private readonly MyCacheService _cache;
    
    public AddCacheMethodsToJavaScript(MyCacheService cache)
    {
       _cache = cache;
    }

    public Task Handle(EvaluatingJavaScriptExpression notification, CancellationToken cancellationToken)
    {
        var engine = notification.Engine;
        engine.SetValue("cache", _cache);
        return Task.CompletedTask;
    }
}
```

**Comment by mircotamburini** [comment, marked as answer]
I implemented with Javascript function 

public class CustomJavaScriptHandler : INotificationHandler<EvaluatingJavaScriptExpression>
{
        private readonly IDistributedCache _cache;
        private readonly ILogger _logger;
        public FetchJavaScriptHandler(ILogger<FetchJavaScriptHandler> logger, IDistributedCache cache)
        {
            _logger = logger;
            _cache = cache;

        }

        public Task Handle(EvaluatingJavaScriptExpression notification, CancellationToken cancellationToken)
        {
            var engine = notification.Engine;
            //Global variable.
            engine.SetValue("getGlobalVariable", (Func<string, string>)(CacheGet));
            engine.SetValue("setGlobalVariable", (Func<string, string, bool>)(CacheSet));
        }

        private string CacheGet(string key)
        {
            try
            {
                var v = _cache.GetString(key);
                return v;
            }
            catch (Exception ex)
            {

                return null;
            }

        }

        private bool CacheSet(string key, string value)
        {
            try
            {
                var options = new DistributedCacheEntryOptions
                {
                    // 24 hours expiring (todo configure)
                    SlidingExpiration = TimeSpan.FromSeconds(60 * 60 * 24)
                };

                _cache.SetString(key, value, options);
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, ex.Message);
                return false;
            }

        }
}

**Comment by mircotamburini** [comment]
Damn.... Sorry i didnt received a answer so i implemented a similar solution 
Thx

**Reply by sfmskywalker** [reply]
No worries, I'm glad to see you figured it out 😃
